usr/src/uts/common/fs/zfs/dmu.c

```diff
@@ -1476,6 +1476,44 @@ dmu_return_arcbuf(arc_buf_t *buf)
  * dmu_write().
  */
 void
+dmu_assign_arcbuf_dnode(dnode_t *dn, uint64_t offset, arc_buf_t *buf,
+    dmu_tx_t *tx)
```

Comment:
It is a copy and paste. The only reason I duplicated it because I was not sure whether it was a good idea to make the entire function into a DB_DNODE_ENTER/DB_DNODE_EXIT critical section for existing callers. If you are fine with that, it should be a trivial change to eliminate the code duplication.